### PR TITLE
Feature/reservation scrapper

### LINKED ISSUE #35: Scrapper for book reservations

### PR DESCRIPTION
**Pull request by monkin77**

### CODE REVIEW

File: src/app_feup/lib/redux/action_creators.dart [hunk 1]
```diff
@@ -261,6 +264,27 @@ ThunkAction<AppState> setBookSearchFilters(SearchFilters filters) {
...
+  return (Store<AppState> store) async {
+    try {
+      store.dispatch(SetCatalogReservationsStatus(RequestStatus.busy));
+
+      final Set<BookReservation> reservations = await library.getReservations();
+
+      store.dispatch(SetCatalogReservations(reservations.toList()));
+      store.dispatch(SetCatalogReservationsStatus(RequestStatus.successful));
+    } catch (e) {
+      Logger().e(e.toString());
```
Comment (laureansvictor) [inline src/app_feup/lib/redux/action_creators.dart:280]: I would just suggest to remove some Logger() from the main branch.

**Review by laureansvictor — APPROVED** [review]

File: src/app_feup/lib/redux/action_creators.dart (hunk 1 above)
Comment (ruialves35) [reply src/app_feup/lib/redux/action_creators.dart:280]: This log is from errors so I think it's worth since they were already doing that

**Review by laureansvictor — APPROVED** [review]

File: src/app_feup/lib/controller/library/library.dart [hunk 2]
```diff
@@ -27,40 +28,75 @@ class Library implements LibraryInterface {
...
 
     return library;
   }
 
+  /**
+   * Wrapper to send a request, after first getting the aleph cookie
+   * pdsCookie is an optional argument
+   */
+  static Future<http.Response> libRequestWithAleph(String url,
+      {Cookie pdsCookie = null}) async {
```
Comment (BrunoRosendo) [inline src/app_feup/lib/controller/library/library.dart]:
```suggestion
      {Cookie pdsCookie}) async {
```
I think this does the same

File: src/app_feup/lib/controller/library/library.dart [hunk 3]
```diff
@@ -27,40 +28,75 @@ class Library implements LibraryInterface {
...
   }
 
-  Future<void> getReservationPage() async {
-    final Cookie alephCookie = await parseAlephCookie();
+  @override
+  Future<Set<BookReservation>> getReservations() async {
+    final reservationHistoryResponse = await libRequestWithAleph(
+        reservationHistoryUrl(this.faculty, this.pdsCookie.value));
+
+    // Logger().i("Reservation request:", reservationHistoryResponse.body);
```
Comment (BrunoRosendo) [inline src/app_feup/lib/controller/library/library.dart]:
```suggestion
```

File: src/app_feup/lib/controller/library/parser_library.dart [hunk 4]
```diff
@@ -57,27 +60,74 @@ class ParserLibrary implements ParserLibraryInterface {
...
+        case 'ano':
+          {
+            bookDetails['year'] = info;
+          }
+          break;
+        case 'isbn':
+          {
+            bookDetails['isbn'] = info;
+          }
+          break;
```
Comment (BrunoRosendo) [inline src/app_feup/lib/controller/library/parser_library.dart]: I feel like all the `{}` here bloater the code but I'll leave the decision up to you

File: src/app_feup/lib/redux/reducers.dart [hunk 5]
```diff
@@ -74,6 +74,14 @@ AppState appReducers(AppState state, dynamic action) {
     return setUserFaculties(state, action);
   } else if (action is SetRestaurantsAction) {
     return setRestaurantsAction(state, action);
+  } else if (action is SetBooksAction) {
+    return setBooks(state, action);
+  } else if (action is SetBooksStatusAction) {
+    return setBooksStatus(state, action);
```
Comment (BrunoRosendo) [inline src/app_feup/lib/redux/reducers.dart]: These two are repeated above. You should remove them

**Review by BrunoRosendo — CHANGES_REQUESTED** [review]
We have the problem that reservations are different from loans but I think this is already big and pretty good, so let's try to merge it

Left some small comments

File: src/app_feup/lib/controller/library/parser_library.dart (hunk 4 above)
Comment (ruialves35) [reply src/app_feup/lib/controller/library/parser_library.dart]: I agree but we searched and switch case were like this :( I'll try without the {} and if it works i take them out 👍

**Comment by ruialves35** [comment]
> We have the problem that reservations are different from loans but I think this is already big and pretty good, so let's try to merge it
> 
> Left some small comments

yep, I think our scrapper wont be hard to extend in order to get that loans. I think we can merge it and then change this in order to get the reservations correctly

**Review by BrunoRosendo — APPROVED** [review]

**Pull request merged**